internal/pahlimiter/limiter_test.go

```diff
@@ -0,0 +1,329 @@
+package pahlimiter
+
+import (
+	"context"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// something "quick" but long enough to not succumb to test noise.
+// non-limited tests should be many, many times faster than this.
+// all tests must be parallel so the total time for the suite is kept low
+var fastTest = 100 * time.Millisecond
+
+func countWhileLeaking(method func(context.Context) (bool, func()), timeout time.Duration) (started int) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	for i := 0; i < 100; i++ {
+		ok, _ := method(ctx) // intentionally leaking close-resources
+		if ok {
+			started += 1
+		}
+	}
+	return started
+}
+
+func TestUnlimited(t *testing.T) {
+	t.Parallel()
+
+	// sharing one across all tests to exercise it further
+	l, err := NewUnlimited()
+	require.NoError(t, err, "should be impossible")
+	defer l.Close()
+
+	t.Run("should allow many history requests", func(t *testing.T) {
+		t.Parallel()
+		started := countWhileLeaking(l.GetHistory, fastTest)
+		assert.Equal(t, 100, started, "all requests should have been allowed")
+	})
+	t.Run("should allow many poll requests", func(t *testing.T) {
+		t.Parallel()
+		started := countWhileLeaking(l.Poll, fastTest)
+		assert.Equal(t, 100, started, "all requests should have been allowed")
+	})
+}
+
+func TestHistoryLimited(t *testing.T) {
+	t.Parallel()
+
+	t.Run("should not limit poll requests", func(t *testing.T) {
+		t.Parallel()
```

Comment:
is the duplicate use of `t.Parallel` here intentional?

Reply:
yep.  the "outer" one allows this top-level test to run concurrently with other top-level tests.  the "inner" ones allow each sub-test to run concurrently with tests inside this test.

both are needed to allow _all_ sub-tests in this file to run simultaneously.  e.g. remove the outer one and `TestHistoryLimited` will start, run all sub-tests, and complete them all before `TestUnlimited` is allowed to unblock its parallel tests.